### ISSUE TITLE
Marking third-party classes as immutable?

### ISSUE THREAD

**Issue by sspat** [vimeo/psalm#2163]
Great feature with the immutability annotations, but I seem to have some problem while trying to use it.
Consider the following code using the `moneyphp/money` library, it's just a wrapper for it:

```php
<?php

declare(strict_types=1);

namespace App\Domain\ValueObject;

use Money\Currency as LibCurrency;
use Money\Money as LibMoney;

/** @psalm-immutable */
class Money
{
    /** @var string */
    private $amount;

    /** @var string */
    private $currency;

    private function __construct(string $amount, string $currency)
    {
        $this->amount   = $amount;
        $this->currency = $currency;
    }

    public function isNegative() : bool
    {
        return $this->getMoneyObject()->isNegative();
    }

    private function getMoneyObject() : LibMoney
    {
        return new LibMoney($this->amount, new LibCurrency($this->currency));
    }
}
```

Psalm is complaining: `ImpureMethodCall - src/Domain/ValueObject/Money.php:194:41 - Cannot call an possibly-mutating method Money\Money::isnegative from a mutation-free context
        return $this->getMoneyObject()->isNegative();`

I can just suppress that error on all methods of the wrapper, but could there be a more elegant solution like marking the third-party class as immutable?
Something like:
```php
<?php

declare(strict_types=1);

namespace App\Domain\ValueObject;

use Money\Currency as LibCurrency;
use Money\Money as LibMoney;

/**
* @psalm-immutable
* @psalm-immutable LibMoney
* @psalm-immutable LibCurrency
*/
class Money
{
    ...
}
```

**Comment by weirdan** [comment]
You could add a stub for Money class, using `<stubs>` config section: https://github.com/vimeo/psalm/blob/9ad6c36d9b15b9fc2181d820a755483e4f355602/docs/running_psalm/configuration.md#stubs

**Comment by muglug** [comment]
Yeah, I'd recommend using stubs

**Comment by muglug** [comment]
Also if you have success with those stubs, you might later encourage those libraries to add `@psalm-immutable` annotations

**Comment by sspat** [comment]
Thanks! Tried out the stubs, it works, but seems somewhat "hacky" to me.
Given an implementation is provided, could this be considered as a feature?
```php
<?php

/**
* @psalm-immutable
* @psalm-immutable SomeClass
*/
class MyClass
{
    /** @var SomeClass */
    private $someClass;
    /** @var SomeOtherClass */
    private $someOtherClass;

    public function someMethod()
    {
        $this->someClass->someMethod();
    }

    /**
    * @psalm-pure SomeOtherClass::someMethod()
    */
    public function someOtherMethod()
    {
        $this->someOtherClass->someMethod();
    }
}
```

**Comment by muglug** [comment]
No - please encourage those 3rd-party packages to use Psalm's annotations, or use stubs.
